src/messaging/Channel.h

```diff
@@ -0,0 +1,248 @@
+
+    ChannelBuilder & SetNetworkPreference(NetworkPreference preference)
+    {
+        mNetworkPreference = preference;
+        return *this;
+    }
+    NetworkPreference GetNetworkPreference() const { return mNetworkPreference; }
+
+    ChannelBuilder & SetTransportPreference(TransportPreference preference)
+    {
+        mTransportPreference = preference;
+        return *this;
+    }
+    TransportPreference GetTransportPreference() const { return mTransportPreference; }
+
+    ChannelBuilder & SetSessionType(SessionType preference)
+    {
+        mSessionType = preference;
+        return *this;
+    }
+    SessionType GetSessionType() const { return mSessionType; }
+
+    uint16_t GetPeerKeyID() const { return mCaseParameters.mPeerKeyId; }
+    ChannelBuilder & SetPeerKeyID(uint16_t keyId)
+    {
+        if (mSessionType == SessionType::kSession_CASE)
+            mCaseParameters.mPeerKeyId = keyId;
+        return *this;
+    }
+
+    uint32_t GetPeerSetUpPINCode() const { return mPaseParameters.mPeerSetUpPINCode; }
+    ChannelBuilder & SetPeerSetUpPINCode(uint32_t peerPINCode)
+    {
+        if (mSessionType == SessionType::kSession_PASE)
+            mPaseParameters.mPeerSetUpPINCode = peerPINCode;
+        return *this;
+    }
+
+#ifndef NDEBUG
+    Optional<Inet::IPAddress> GetHintPeerAddress() const { return mHintPeerAddr; }
+    ChannelBuilder & SetHintPeerAddress(Inet::IPAddress peerAddr)
+    {
+        mHintPeerAddr.SetValue(peerAddr);
+        return *this;
+    }
+#endif // NDEBUG
+
+private:
+    NodeId mPeerNodeId                       = kUndefinedNodeId;
+    NetworkPreference mNetworkPreference     = NetworkPreference::kNetwork_Default;
+    TransportPreference mTransportPreference = TransportPreference::kTransport_Default;
+    SessionType mSessionType                 = SessionType::kSession_Default;
+    union
+    {
+        struct
+        {
+            uint16_t mPeerKeyId;
+        } mCaseParameters;
+        struct
+        {
+            uint32_t mPeerSetUpPINCode;
+        } mPaseParameters;
+    };
+
+#ifndef NDEBUG
+    Optional<Inet::IPAddress> mHintPeerAddr;
+#endif // NDEBUG
+};
+
+class ExchangeContext;
+class ExchangeDelegate;
+
+enum class ChannelState
+{
+    kChanneState_None,
+    kChanneState_Preparing,
+    kChanneState_Ready,
+    kChanneState_Closed,
+    kChanneState_Failed,
+};
+
+class ChannelContextHandleAssociation;
+
+/**
+ *  @brief
+ *    ChannelHandle is a reference to a channel. An active ChannelHandle will
+ *    keep the channel available and ready for use, such that a message can be
+ *    sent immediately to the peer.
+ *
+ *    The ChannelHandle controls the lifespan of the channel. The ChannelHandle
+ *    objects are not copyable, so there can be only one instance referring to
+ *    the channel. When the handler is released, the channel will be flagged as
+ *    pending close, and if there is no active exchange which is using the
+ *    channel, the channel will be closed.
```

Comment:
No!

With current concept. The application can create a channel pass the channel to a protocol, the protocol will create an exchange to do works. At the mean time, the application can free its handle, it wont invalid the exchange inside the protocol. If the application want to terminate the job, it can destroy the protocol.